instrumentation/opentelemetry-instrumentation-openai/src/opentelemetry/instrumentation/openai/patch.py

```diff
@@ -0,0 +1,438 @@
+        SpanAttributes.LLM_IS_STREAMING: kwargs.get("stream"),
+        SpanAttributes.LLM_REQUEST_TEMPERATURE: kwargs.get("temperature"),
+        SpanAttributes.LLM_TOP_K: top_k,
+        SpanAttributes.LLM_PROMPTS: json.dumps(prompts) if prompts else None,
+        SpanAttributes.LLM_USER: user,
+        SpanAttributes.LLM_REQUEST_TOP_P: top_p,
+        SpanAttributes.LLM_REQUEST_MAX_TOKENS: kwargs.get("max_tokens"),
+        SpanAttributes.LLM_SYSTEM_FINGERPRINT: kwargs.get("system_fingerprint"),
+        SpanAttributes.LLM_PRESENCE_PENALTY: kwargs.get("presence_penalty"),
+        SpanAttributes.LLM_FREQUENCY_PENALTY: kwargs.get("frequency_penalty"),
+        SpanAttributes.LLM_REQUEST_SEED: kwargs.get("seed"),
+        SpanAttributes.LLM_TOOLS: json.dumps(tools) if tools else None,
+        SpanAttributes.LLM_TOOL_CHOICE: kwargs.get("tool_choice"),
+        SpanAttributes.LLM_REQUEST_LOGPROPS: kwargs.get("logprobs"),
+        SpanAttributes.LLM_REQUEST_LOGITBIAS: kwargs.get("logit_bias"),
+        SpanAttributes.LLM_REQUEST_TOP_LOGPROPS: kwargs.get("top_logprobs"),
+    }
+
+
+class StreamWrapper:
+    span: Span
+
+    def __init__(
+        self, stream, span, prompt_tokens, function_call=False, tool_calls=False
+    ):
+        self.stream = stream
+        self.span = span
+        self.prompt_tokens = prompt_tokens
+        self.function_call = function_call
+        self.tool_calls = tool_calls
+        self.result_content = []
+        self.completion_tokens = 0
+        self._span_started = False
+        self.setup()
+
+    def setup(self):
+        if not self._span_started:
+            self.span.add_event(Event.STREAM_START.value)
+            self._span_started = True
+
+    def cleanup(self):
+        if self._span_started:
+            self.span.add_event(Event.STREAM_END.value)
+            set_span_attribute(
+                self.span,
+                SpanAttributes.LLM_USAGE_PROMPT_TOKENS,
+                self.prompt_tokens,
+            )
+            set_span_attribute(
+                self.span,
+                SpanAttributes.LLM_USAGE_COMPLETION_TOKENS,
+                self.completion_tokens,
+            )
+            set_span_attribute(
+                self.span,
+                SpanAttributes.LLM_USAGE_TOTAL_TOKENS,
+                self.prompt_tokens + self.completion_tokens,
+            )
+            set_event_completion(
+                self.span,
+                [
+                    {
+                        "role": "assistant",
+                        "content": "".join(self.result_content),
+                    }
+                ],
+            )
+
+            self.span.set_status(StatusCode.OK)
+            self.span.end()
+            self._span_started = False
+
+    def __enter__(self):
+        self.setup()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.cleanup()
+
+    async def __aenter__(self):
```

Comment:
I see you are implementing some async methods but you are not patching the AsyncClient. Will it be added later?

Reply:
i agree with that, we can start with a smaller scope and just add async support in the future